spaceship/views.py

```diff
@@ -418,92 +455,153 @@ def roster(team_id):
 
 @app.route('/enlist/<key>', methods=['GET', 'POST'])
 def enlist(key):
-  invitation = (Invitation
-                .select()
-                .where(Invitation.key_for_sharing == key)
-                .get())
+  try:
+    invitation = (Invitation
+                  .select()
+                  .where(Invitation.key_for_sharing == key)
+                  .get())
+  except Invitation.DoesNotExist:
+    invitation = None
   if not invitation:
     flash({'msg':f'Could not find invitation', 'level':'danger'})
     return redirect(url_for('home'))
   if invitation.status == 'accepted':
-    return redirect(url_for('dashboard'))
+    flash({'msg':f'Invitation was already accepted.', 'level':'danger'})
+    return redirect(url_for('home'))
+
+  decline = DeclineInvitation()
+  if decline.decline.data and decline.validate():
+    with db.atomic() as transaction:
+      try:
+        invitation.status = 'declined'
+        invitation.save()
+      except DatabaseError:
+        transaction.rollback()
+        flash({'msg':f'Database error', 'level':'danger'})
+      else:
+        flash({'msg':f'Invitation declined', 'level':'success'})
+    return redirect(url_for('home'))
+
+  register = None
 
   if current_user.is_authenticated:
-    enlist = EnlistExistingUser()
     email_mismatch = (invitation.invited_email and current_user.email != invitation.invited_email)
-    if not enlist.is_submitted() and email_mismatch:
+    if email_mismatch:
       # could have gotten the invite via an alias? but also maybe multiple accounts. let them decide
       flash({'msg':f'Invitation was for ' + invitation.invited_email + ' but you are logged in as ' + current_user.email, 'level':'warning'})
   elif User.select().where(User.email == invitation.invited_email):
     # assume cookies were cleared
     flash({'msg':f'Please log in as ' + invitation.invited_email, 'level':'warning'})
     return redirect(url_for('login', next=url_for('enlist', key=key)))
   else:
-    enlist = EnlistNewUser()
-    if not enlist.is_submitted():
+    register = Register()
+    if not register.is_submitted():
       # auto populate the email from the invitation but allow the user to change it
-      enlist.email.process_data(invitation.invited_email)
+      register.email.process_data(invitation.invited_email)
+
+  oauth_user_info = session.pop('oauth_user_info', None)
+  session['oauth_next_url'] = url_for('enlist', key=invitation.key_for_sharing)
+
+  name = ''
+  email_address = ''
+  password = None
+  if oauth_user_info:
+    name = oauth_user_info.get('name', '')
+    email_address = oauth_user_info.get('email', '')
+  elif register and register.validate_on_submit():
+    name = register.data['name']
+    email_address = register.data['email']
+    password = register.data['password']
+
+  if name and email_address:
+    with db.atomic() as transaction:
```

Comment:
mmm, we have a bit of copy-pasta. i think our views are starting to get a little heavy, i'd like to refactor some of this code into [service objects](https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1) or some equivalent way to encapsulate business logic (since service objects seem very RoR-oriented)

Reply:
hrm, yeah, we could just move related sets of db mutation and queries out into functions in logically named files. i'll take a look at this after some ui tweaks and other stuff for mvp is in.

Reply:
tracked in issue #52